### PR TITLE
Fixed #30672 -- Fixed crash of JSONField/HStoreField key transforms on expressions with params.

### PR DESCRIPTION
**Pull request by felixxm**

### CODE REVIEW

File: django/contrib/postgres/fields/array.py [hunk 1]
```diff
@@ -262,7 +262,7 @@ def __init__(self, index, base_field, *args, **kwargs):
 
     def as_sql(self, compiler, connection):
         lhs, params = compiler.compile(self.lhs)
-        return '%s[%s]' % (lhs, self.index), params
+        return '%s[%%s]' % lhs, [self.index] + params
```
Comment (apollo13) [inline django/contrib/postgres/fields/array.py]: Shouldn't that have params and index swapped? I fear we also might have to fix this in the existing released patches?

Comment (apollo13) [reply django/contrib/postgres/fields/array.py]: If the answer to my previous question is yes the we also need a new test for that, apparently we nowhere tested with extra params :D

Comment (felixxm) [reply django/contrib/postgres/fields/array.py]: Currently both can be used only with field name, i.e. `params` are always an empty array `[]` so it's not an issue.

Comment (felixxm) [reply django/contrib/postgres/fields/array.py]: Swapped.

**Comment by felixxm** [comment]
@apollo13 I swapped params with key/index in `JSONField`/`HStoreField` key and index transforms.

**Comment by apollo13** [comment]
@felixxm Could you also add a simple test which just initializes the Transform with a lhs which already has a param so we can check the outputs?

**Comment by felixxm** [comment]
@apollo13 I added tests.

**Comment by felixxm** [comment]
I added release notes and rebased from master.

**Comment by felixxm** [comment]
@apollo13 @carltongibson Thanks for reviews :+1:

**Pull request merged**